jekyll/_posts/2018-06-03-create-a-simple-macro.md

```diff
@@ -0,0 +1,120 @@
+
+---
+title: "Create a simple macro"
+author: Antonis Geralis
+---
+
+Hello, as you might know Nim is a powerful programming language that supports
+metaprogramming using macros. Though a lot of Nim programmers are unaware about
+the merits of them due to lack of learning resources. In the first part of
+this series we will discuss the use of macros to simplify the creation of
+boilerplate code in Nim.
+
+Suppose we have code which builds a directed graph.
+
+```nim
+  proc buildCityGraph(): Digraph =
+    result = initGraph()
+    result.addEdge(initEdge(initNode("Boston"), initNode("Providence")))
+    result.addEdge(initEdge(initNode("Boston"), initNode("New York")))
+    result.addEdge(initEdge(initNode("Providence"), initNode("Boston")))
+    result.addEdge(initEdge(initNode("Providence"), initNode("New York")))
+    ...
+```
+
+A template could be used to reduce the amount of typing. For example:
+
+```nim
+   template adder(graph, src, dest): untyped =
+      graph.addEdge(initEdge(initNode(src), initNode(dest)))
+```
+
+However it needs to be called each time with all three arguments. This excludes
+the possibility of using an operator with a nice syntax. Although operators can
+have more than two parameters, we are limited into calling them like
+this: `` `->`(result, "Boston", "Providence") ``. A macro, however can be used
+to create a simpler and more appealling syntax.
+
+> Macros can be used to implement domain specific languages.
+[*From the Nim manual.*](https://nim-lang.org/docs/manual.html#macros)
+
+To begin, this is how calling our macro will look like:
+
+```nim
+  proc buildCityGraph(): Digraph =
+    result = initGraph()
```

Comment:
Wouldn't it be better to generate this in the macro?

Reply:
You mean the init function? Well I thought (also sections like edges: ...) so but I just wanted something really simple.